### PR TITLE
feat(cli): build minimal theme family artifacts

### PR DESCRIPTION
**Pull request by cixzhang**
## User need

Theme package authors who maintain a base theme and related descendants need one complete stylesheet they can load before first paint, then switch by theme identity without loading or ordering member files. Standalone builds remain the simpler fallback when an app needs only one theme.

## Current specification

[AST-034: Theme family build behavior](https://github.com/facebook/astryx/blob/8e5e42637ef2c911b980fb27a7108867ce4c7635/docs/specs/AST-034/spec.md) is the current contract for one deterministic keyed CSS/JS/declaration trio, complete selected lineage, zero-delta identity, attribute-only switching, and unchanged standalone builds. Prefix-independent local-token behavior is on `main` through [PR #6285](https://github.com/facebook/astryx/pull/6285); this PR is based directly on that merged result.

## Public delta

- `astryx theme build --family <base> <descendants...> --family-key <key>` emits exactly `<key>.css`, `<key>.js`, and `<key>.d.ts` beside the root source. `--family-key` changes only the filename stem.
- One CSS file contains every selected member. Shared declarations are grouped under zero-specificity family identity scopes while their inner selectors retain standalone specificity; descendant differences use member scopes. A zero-delta descendant remains selectable and exported without a per-member file.
- One CSS-free ESM module exports every complete resolved member; one declaration file describes those exports and includes their existing custom-value augmentations. `--check` regenerates and compares that exact trio through the existing theme-build semantics.
- Family mode composes with `--check` and existing `--icons-specifier` behavior, and is refused with `--watch` or `--out`. Ordinary single- and multi-theme builds remain unchanged.

## Design and alternatives

This replaces the prior 50-file transactional draft with thin orchestration over the existing theme loader, Core rule generators, serializers, check comparison, and staged writer. Core runtime themes and types are byte-identical to `main`: family ancestry is captured only inside the CLI loader, retained privately across spread/frozen/cached exports, and stripped before output. Selected sources and intervening static import/re-export modules load dependency-first through one intercepted cache; any path whose exact ancestry cannot be observed fails closed.

One family helper validates and orders the graph, then compares existing generated declarations and sections without inventing a second compiler or artifact model. Computed or otherwise non-static registry dependencies are rejected before writes so accepted registry graphs are provably resolvable and CSS-free.

Rejected here: per-member family files; consumer-managed ancestor order; and a parallel plan/manifest/transaction/registry backend. The generated CSS is eagerly downloaded so the complete family is ready on first paint; consumers needing only one member keep the standalone build.

Before → after implementation size:

- 50 files, +6,786/−283 → 22 files, +2,227/−165
- production: 4 files, +1,184/−135
- tests and permanent Chromium evidence: focused structural, compatibility, and cascade coverage
- docs/release note: 4 files, +84/−10
- Sandbox consumer integration: 10 files, +304/−12
- CI: the existing `theme-layers` job is extended; no new job or required status

## Evidence

- Focused family contract: 9/9 tests, including deterministic shuffles, default/raw/spread/TLA/frozen/cached exact ancestry, an unselected re-export barrel, shared/delta CSS, zero-delta identity, recursive CSS-free static registry graphs, fail-closed computed dependencies and unobservable ancestry, exact-source impostor rejection, `--check`, and standalone preservation.
- Existing standalone/theme-build compatibility: 237/237 tests, including packed older-Core coverage.
- Existing Core theme compatibility: 219/219 tests; `defineTheme` source/tests and ordinary runtime keys, symbols, spread, and JSON behavior are unchanged from merged base `ae9ccef67c09a036c1f2e273e783ae78cd8d2c67`.
- Sandbox tests: 22/22. Storybook tests: 19 passed, 5 skipped.
- Representative standalone generated CSS, JS, and declarations remain byte-identical to the merged base.
- Strict CLI/Core/Storybook/Sandbox typechecks, repository checks, docs drift, copyright, lint, family ESM import, generated declaration consumer typecheck, root build, Storybook build, and the normal Sandbox generate/build path pass.
- Final review findings are fixed at `a843584259500b81f0e3c4d62e70dd46f4e7eb77`: the response preserves the root theme identity, staged write failure cannot report or leave a mixed trio, and the maintained family matrix now has permanent Chromium assertions.
- The permanent Sandbox Themes page consumes the actual CLI-generated CSS/ESM from the normal Sandbox build, while leaving the generated trio ignored and uncommitted. It shows every identity, attribute-only switching, nested and sibling roots, the zero-delta descendant, representative component/prose/adaptation/on-media behavior, and equal-specificity same-layer CSS before and after the family.
- Exact-head CI [34921935320](https://github.com/facebook/astryx/actions/runs/34921935320) passes every applicable check, including the normal Sandbox build, the extended `theme-layers` Chromium guard, `pr-a11y`, RTL shards, smoke, registry, lint, and visual acceptance. Stable visual regression correctly reports no stable visual scope. This PR adds no new job or required status.

## Scope

- [x] One capability is added.
- [x] Unrelated fixes, cleanup, and policy changes were removed or split.
- [x] Consumer docs, focused tests, and migration evidence ship with the feature.
- [x] Public text and artifacts contain no internal Meta context.

## Testing

```sh
pnpm vitest run --project node packages/cli/api/theme/build packages/cli/clients/cli/commands/build-theme*.test.mjs --maxWorkers=2
pnpm vitest run --project ui packages/core/src/theme/defineTheme.test.ts packages/core/src/theme/generateThemeRules.test.ts packages/core/src/theme/themeAdaptations.test.ts --maxWorkers=2
pnpm -F @astryxdesign/sandbox test
pnpm -F @astryxdesign/sandbox build
pnpm -F @astryxdesign/storybook build
pnpm -F @astryxdesign/cli typecheck:strict
pnpm -F @astryxdesign/core typecheck
pnpm check:repo
```

The Sandbox page is the maintained browser surface. It loads only the generated family stylesheet and uses the generated ESM identities; compiler edge cases remain in temp-generated tests. No bespoke browser harness or matrix remains in the change.

### CODE REVIEW

File: packages/cli/api/theme/build/family/transaction.mjs [hunk 1]
```diff
@@ -0,0 +1,990 @@
...
+      } catch (error) {
+        const previousDirectory = path.join(
+          root,
+          ...journal.prev.target.split('/'),
+        );
+        const previousStat = fs.lstatSync(previousDirectory, {
+          throwIfNoEntry: false,
+        });
+        if (
+          !previousStat?.isDirectory() ||
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/family/transaction.mjs]: Fixed in [`86545f1`](https://github.com/facebook/astryx/commit/86545f114ab5b37f5e1d10a63b4b24c6b5d3856d): post-switch recovery now treats an absent predecessor as already cleaned while still rejecting any surviving nonempty or invalid state. The new fault-injection test covers the post-cleanup crash boundary, repeated recovery, `--check`, and a later publish. Focused family tests pass 88/88, and [exact-head CI](https://github.com/facebook/astryx/actions/runs/34700774286) is green.

File: packages/cli/api/theme/build/family/transaction.mjs [hunk 2]
```diff
@@ -0,0 +1,990 @@
...
+    ]);
+    for (const relative of priorOwned) checked.add(relative);
+
+    for (const [relative, content] of files) {
+      const file = path.join(
+        root,
+        ...current.target.split('/'),
+        ...relative.split('/'),
+      );
+      if (!fs.existsSync(file)) {
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/family/transaction.mjs]: Fixed in [`86545f1`](https://github.com/facebook/astryx/commit/86545f114ab5b37f5e1d10a63b4b24c6b5d3856d): `--check` now sends every expected owned path through the existing `lstat` parent/final regular-file guard before reading bytes. Matching-byte external and internal symlinks are reported outdated; the same guard rejects non-regular entries and escaping paths. Both symlink regressions pass, the focused family suite is 88/88, and [exact-head CI](https://github.com/facebook/astryx/actions/runs/34700774286) (including `theme-layers`) is green.

File: packages/cli/api/theme/build/build.mjs [hunk 3]
```diff
@@ -2406,3 +2500,300 @@ Or with a <link> tag:
...
+      const inheritedIndicators = parentBinding
+        ? `${parentBinding}.indicators`
+        : null;
+      /** @param {string | null} own @param {string | null} inherited */
+      const expression = (own, inherited) =>
+        own && inherited
+          ? `{...${inherited}, ...${own}}`
+          : (own ?? inherited ?? undefined);
+      const indicatorImport = indicator.info
+        ? `import { ${indicator.info.exportName} as ${ownIndicators} } from ${JSON.stringify(indicator.specifier ?? indicator.info.importPath)};\n`
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/build.mjs:2850]: Fixed in `7951ec5`. Before family publication or `--check` success, emitted registry imports are resolved from the aggregate module, known local and package dependency edges are followed recursively, imported and re-exported bindings are validated, and CSS, missing, extensionless, or non-static local edges fail with zero writes. Unknown package loaders remain permitted while known package edges are still inspected; CommonJS uncertainty is runtime-validated. Focused tests cover valid nested, bare-package, and CommonJS graphs plus each failure case. Both exact-head reviews are clean, and the full local compatibility gate passes.

File: packages/cli/api/theme/build/family.mjs [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+    }
+    const parentName = parent?.name ?? null;
+    if (parentName && !byName.has(parentName)) {
+      throw new Error(
+        `Theme "${member.theme.name}" is missing selected ancestor "${parentName}".`,
+      );
+    }
+    if (
+      parentName &&
+      fingerprint(parent) !== fingerprint(byName.get(parentName).theme)
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/family.mjs]: Fixed in `7951ec5`. All selected members are prepared through one shared cached Jiti loader, the exact loaded `sourceTheme` object is retained, and each lineage edge must satisfy `parent === selected.sourceTheme`. A focused counterexample using a structurally identical but separately loaded, unselected parent now rejects. Both exact-head reviews are clean.

File: .github/scripts/theme-layer-cascade.js [hunk 5]
```diff
@@ -123,15 +135,259 @@ async function buildFixture(outDir) {
...
+      beforeOrder,
+      'physical order changed the scoped winner',
+    );
+    const familySnapshots = new Map();
+    for (const name of names)
+      familySnapshots.set(name, await familySnapshot(page, name));
+    const calm = familySnapshots.get('ocean-calm');
+    // prettier-ignore
+    checks([
+      ['inherited hover', calm.hoverBackground, 'rgb(220, 40, 40)', 'resting rule suppressed hover'],
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply .github/scripts/theme-layer-cascade.js]: Fixed in `7951ec5`. The browser fixture now uses a dedicated unvarianted `#hover-button`, so the inherited base `:hover` rule is no longer masked by the primary variant. The guard compares that same red hover state in family and standalone output on both native-link and Vite paths. Both exact-head reviews are clean; the real Chromium lane is rerunning on this pushed head.

File: packages/cli/api/theme/build/build.mjs [hunk 6]
```diff
@@ -1136,7 +1185,85 @@ const THEME_MODULE_EXTENSIONS = [
...
+  while (pending.length > 0) {
+    const current = pending.pop(); if (!current) break; const {specifier, parent, commonJs, imports, root} = current;
+    if (/\.css(?:$|[?#])/i.test(specifier)) throw registryGraphError(`Theme family registry graph imports CSS: ${specifier}`);
+    const file = resolveRegistryModule(specifier, parent, commonJs); if (!file) continue;
+    if (/\.css$/i.test(file)) throw registryGraphError(`Theme family registry graph imports CSS: ${file}`);
+    const external = file.split(path.sep).includes('node_modules');
+    for (const name of imports) if (external) runtimeChecks.push({file, exportName: name}); else { const named = hasNamedExport(file, name); if (named === false) throw registryGraphError(`Family registry ${file} does not export "${name}".`); if (named === null) runtimeChecks.push({file, exportName: name}); }
+    if (root) runtimeChecks.push({file, exportName: root.exportName});
+    if (seen.has(file)) continue; seen.add(file);
+    if (/\.[cm]?[jt]sx?$/i.test(file)) for (const edge of moduleDependencies(fs.readFileSync(file, 'utf8'), file, !external)) pending.push({...edge, parent: file});
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/build.mjs]: Fixed in `053de6b`. Registry traversal is now fail-closed for every non-static `import()` or `require()` in local and external dependency graphs, with the precise error that the graph cannot be proven resolvable and CSS-free. A temp external package reproduces computed `import(`./hidden.${extension}`)` with a real hidden CSS file; family build rejects it with zero trio writes. Accepted aggregate ESM is parsed to prove every executable dependency is static and non-CSS, then imported natively. Focused 7/7, full 235/235, both exact-head reviews clean.

File: packages/core/src/theme/defineTheme.ts [hunk 7]
```diff
@@ -430,6 +430,8 @@ export interface DefinedTheme {
   icons?: ThemeIconOverrides;
   /** Indicator overrides for stateful control visuals, keyed by name */
   indicators?: IndicatorRegistry;
+  /** The exact resolved base retained for build-time family graph discovery. @internal */
+  __extends?: DefinedTheme;
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/core/src/theme/defineTheme.ts]: Fixed in `053de6b`. The public `DefinedTheme.__extends` field and type change are removed. Exact ancestry now travels through a symbol/WeakMap side channel: Core exposes no field or export, the symbol is absent from public iteration/JSON, and the CLI retains exact parents privately then strips all metadata before CSS/ESM/types. Regressions cover default exports, raw inputs, spread exports, frozen themes, cached `--check`, and native `.mjs` top-level await; exact impostor parents still reject. Generated artifacts contain no lineage metadata. Both exact-head reviews clean.

File: apps/sandbox/package.json [hunk 8]
```diff
@@ -5,12 +5,13 @@
   "scripts": {
     "test": "vitest run --root ../.. apps/sandbox --project node",
     "build:deps": "pnpm -F @astryxdesign/theme-neutral build && pnpm -F @astryxdesign/theme-matcha build && pnpm -F @astryxdesign/theme-stone build && pnpm -F @astryxdesign/theme-gothic build && pnpm -F @astryxdesign/theme-chocolate build && pnpm -F @astryxdesign/theme-y2k build && pnpm -F @astryxdesign/theme-butter build",
+    "generate:theme-family": "pnpm exec astryx theme build --family src/themes/theme-family/ocean.ts src/themes/theme-family/ocean-calm.ts src/themes/theme-family/ocean-calm-deep.ts src/themes/theme-family/ocean-midnight.ts --family-key sandbox-ocean-family",
```
Comment (cixzhang) [inline apps/sandbox/package.json:8]: Instead of adding a test ocean theme can we expand on the probe theme and add just 1 family member there to test. Keep our testing scaffolds as minimal as possible.

File: packages/core/src/theme/defineTheme.ts [hunk 9]
```diff
@@ -634,6 +637,12 @@ export function defineTheme(input: DefineThemeInput): ResolvedDefinedTheme {
     __axes,
   };
 
+  // Enumerable symbols survive spread but stay outside public types and JSON.
+  if (base) {
+    // prettier-ignore
+    Object.defineProperty(theme, THEME_PARENT, {value: base, enumerable: true, configurable: true});
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/core/src/theme/defineTheme.ts]: Fixed in adb9cbc0c0ab771e12deb99ed79e9dfc8e633ea8. Core defineTheme and its tests are now byte-identical to main, so ordinary themes regain their exact own keys, symbols, spread behavior, and JSON shape. Exact ancestry now lives only in core-interception.mjs through the CLI loader and WeakMap retention; TLA uses one dependency-first intercepted cache, including unselected static re-export barrels, while unobservable ancestry fails closed. Focused regressions cover default/raw/spread/TLA/frozen/cached check, bypass rejection, impostor rejection, and ordinary Core object shape. The 8 focused tests, 237 build tests, 219 Core tests, all builds/typechecks/lint, and both exact-head reviews pass.

File: packages/cli/api/theme/build/build.mjs [hunk 10]
```diff
@@ -2406,3 +2578,533 @@ Or with a <link> tag:
...
+    import('./family/manifest.mjs'),
+    import('./family/plan.mjs'),
+    import('./family/registries.mjs'),
+    import('./family/transaction.mjs'),
+  ]);
+
+  // Recover every selected source directory that carries a journal for this
+  // exact artifact key before loading any source. This stays invocation-order
+  // independent while avoiding writes to unrelated same-named directories.
+  const recoveryRoots = new Set(
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/build.mjs]: Obsolete after the minimal rewrite: the old recovery/backend path was removed. Family output now validates the selected graph before using the existing staged writer, with no same-key recovery or family transaction backend. No code change is needed on the removed path.

File: packages/cli/api/theme/build/family/esm.mjs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        write: false,
+        outfile: 'registry-check.js',
+        format: 'esm',
+        platform: 'browser',
+        target: 'es2022',
+        treeShaking: false,
+        legalComments: 'none',
+        logLevel: 'silent',
+      });
+    } catch {
```
[the remark this answers is not in the document]
Comment (cixzhang) [reply packages/cli/api/theme/build/family/esm.mjs]: Obsolete after the minimal rewrite: family/esm.mjs and the old registry backend were removed. The current build validates recursive static registry dependencies in build.mjs and rejects computed or otherwise non-static dependencies before any write. No code change is needed on the removed path.
